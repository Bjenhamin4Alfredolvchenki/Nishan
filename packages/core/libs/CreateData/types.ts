import { ICache } from '@nishans/cache';
import { IOperation, TTextFormat } from '@nishans/types';
import { Logger } from '../../types';

export interface ParentCollectionData {
	parent_collection_id: string;
	name: TTextFormat;
	token: string;
	logger?: Logger;
	stack: IOperation[];
	cache: ICache;
	parent_relation_schema_unit_id: string;
}
