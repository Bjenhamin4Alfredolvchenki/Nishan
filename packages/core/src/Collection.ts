import { generateFormulaAST } from '@nishans/notion-formula';
import { ICollection, IPage, TCollectionBlock, TSchemaUnit } from '@nishans/types';
import { FilterType, FilterTypes, ICollectionUpdateInput, IPageCreateInput, IPageUpdateInput, ISchemaMapValue, ISchemaUnitMap, NishanArg, TCollectionUpdateKeys, TSchemaUnitInput, UpdateType, UpdateTypes } from '../types';
import { createSchemaUnitMap, createShortId, getSchemaMap, nestedContentPopulate, Operation, transformToMultiple, warn } from '../utils';
import Data from "./Data";
import Page from './Page';
import SchemaUnit from "./SchemaUnit";

/**
 * A class to represent collection of Notion
 * @noInheritDoc
 */
class Collection extends Data<ICollection> {

  constructor(args: NishanArg) {
    super({ ...args, type: "collection" });
  }

  #getRowPages = async () => {
    await this.initializeNotionCache();
    const page_ids: string[] = [];
    for (const [_, page] of this.cache.block)
      if (page?.type === "page" && page.parent_id === this.id && !page.is_template) page_ids.push(page.id);
    return page_ids;
  }

  getCachedParentData() {
    return this.cache.block.get(this.getCachedData().parent_id) as TCollectionBlock;
  }

  /**
   * Update the collection
   * @param opt `CollectionUpdateParam`
   */
  update(opt: ICollectionUpdateInput) {
    this.updateCacheLocally(opt, TCollectionUpdateKeys)
  }

  /**
   * Create multiple templates for the collection
   * @param opts Array of Objects for configuring template options
   */
  async createTemplates(rows: IPageCreateInput[]) {
    return await nestedContentPopulate(rows, this.id, this.type as "collection", this.getProps())
  }

  /**
   * Get a single template page of the collection
   * @param args string id or a predicate function
   * @returns Template page object
   */
  async getTemplate(args?: FilterType<IPage>) {
    return (await this.getTemplates(typeof args === "string" ? [args] : args, false))[0]
  }

  async getTemplates(args?: FilterTypes<IPage>, multiple?: boolean) {
    return await this.getIterate<IPage, Page[]>(args, {
      child_ids: "template_pages",
      multiple,
      child_type: "block",
      container: []
    }, (page_id) => this.cache.block.get(page_id) as IPage, (id,_,pages) => pages.push(new Page({ ...this.getProps(), id })));
  }

  async updateTemplate(args: UpdateType<IPage, IPageUpdateInput>) {
    return (await this.updateTemplates(typeof args === "function" ? args : [args], false))[0]
  }

  // ? FEAT:1:H Custom update to take care of new properties, using the name rather than the id
  async updateTemplates(args: UpdateTypes<IPage, IPageUpdateInput>, multiple?: boolean) {
    return await this.updateIterate<IPage, IPageUpdateInput, Page[]>(args, {
      child_ids: "template_pages",
      multiple,
      child_type: "block",
      container: [],
    }, (child_id) => this.cache.block.get(child_id) as IPage, 
      (id,_,__,pages) => pages.push(new Page({ ...this.getProps(), id })));
  }

  /**
   * Delete a single template page from the collection
   * @param args string id or a predicate function
   */
  async deleteTemplate(args?: FilterType<IPage>) {
    return await this.deleteTemplates(typeof args === "string" ? [args] : args, false);
  }

  /**
   * Delete multiple template pages from the collection
   * @param args string of ids or a predicate function
   * @param multiple whether multiple or single item is targeted
   */
  async deleteTemplates(args?: FilterTypes<IPage>, multiple?: boolean) {
    await this.deleteIterate<IPage>(args, {
      multiple,
      child_ids: "template_pages",
      child_type: "block",
      child_path: "template_pages",
      container: []
    }, (child_id) => this.cache.block.get(child_id) as IPage)
  }

  /**
   * Add rows of data to the collection block
   * @param rows
   * @returns An array of newly created page objects
   */
  async createPages(rows: IPageCreateInput[]) {
    return await nestedContentPopulate(rows, this.id, this.type as "collection", this.getProps())
  }

  async getPage(arg?: FilterType<IPage>) {
    return (await this.getPages(transformToMultiple(arg), false))[0]
  }

  async getPages(args?: FilterTypes<IPage>, multiple?: boolean) {
    return await this.getIterate<IPage, Page[]>(args, {
      child_ids: await this.#getRowPages(),
      child_type: "block",
      multiple,
      container: []
    }, (id) => this.cache.block.get(id) as IPage, (id,_,pages) => pages.push(new Page({ ...this.getProps(), id })));
  }

  async updatePage(args: UpdateType<IPage, IPageUpdateInput>) {
    return (await this.updatePages(typeof args === "function" ? args : [args], false))[0]
  }

  async updatePages(args: UpdateTypes<IPage, IPageUpdateInput>, multiple?: boolean) {
    return await this.updateIterate<IPage, IPageUpdateInput, Page[]>(args, {
      child_ids: await this.#getRowPages(),
      multiple,
      child_type: "block",
      container: [],
    }, (child_id) => this.cache.block.get(child_id) as IPage, (id,_,__,pages) => pages.push(new Page({ ...this.getProps(), id })));
  }

  async deletePage(args?: FilterType<IPage>) {
    return await this.deletePages(typeof args === "string" ? [args] : args, false);
  }

  /**
   * Delete multiple template pages from the collection
   * @param args string of ids or a predicate function
   * @param multiple whether multiple or single item is targeted
   */
  async deletePages(args?: FilterTypes<IPage>, multiple?: boolean) {
    // ! Push to operation stack and update pages
    await this.deleteIterate<IPage>(args, {
      child_ids: await this.#getRowPages(),
      child_type: "block",
      multiple,
      container: [],
      manual: true
    }, (child_id) => this.cache.block.get(child_id) as IPage, (child_id, child_data)=>{
      child_data.alive = false;
      child_data.last_edited_by_id = this.user_id;
      child_data.last_edited_by_table = "notion_user";
      child_data.last_edited_time = Date.now();
      this.stack.push(Operation.block.update(child_id, [], { alive: false , last_edited_time: Date.now(), last_edited_by: this.user_id}));
    });
  }

  /**
   * Create multiple new columns in the collection schema
   * @param args array of Schema creation properties
   * @returns An array of SchemaUnit objects representing the columns
   */
  async createSchemaUnits(args: TSchemaUnitInput[]) {
    const schema_unit_map = createSchemaUnitMap(), data = this.getCachedData();
    for (let index = 0; index < args.length; index++) {
      const arg = args[index], schema_id = arg.type === "title" ? "title" : createShortId();
      if (!data.schema[schema_id]) {
        const schema_map = getSchemaMap(data.schema);
        if(arg.type === "formula") data.schema[schema_id] = {...arg, formula: generateFormulaAST(arg.formula[0] as any, arg.formula[1] as any, schema_map)}
        // ! Fix this
        // else if(arg.type === "relation") data.schema[schema_id] = generateRelationSchema(arg);
        const schema_obj = new SchemaUnit({ schema_id, ...this.getProps(), id: this.id })
        schema_unit_map[arg.type].set(schema_id, schema_obj);
        schema_unit_map[arg.type].set(arg.name, schema_obj);
        this.logger && this.logger("CREATE", "collection", schema_id);
      } else
        warn(`Collection:${this.id} already contains SchemaUnit:${schema_id}`)
    };

    this.stack.push(Operation.collection.update(this.id, [], { schema: data.schema }));
    this.updateLastEditedProps();
    return schema_unit_map;
  }

  async getSchemaUnit(arg?: FilterType<ISchemaMapValue>) {
    return (await this.getSchemaUnits(transformToMultiple(arg), false))
  }

  /**
   * Return multiple columns from the collection schema
   * @param args schema_id string array or predicate function
   * @returns An array of SchemaUnit objects representing the columns
   */
  async getSchemaUnits(args?: FilterTypes<ISchemaMapValue>, multiple?: boolean) {
    const data = this.getCachedData(), schema_map = getSchemaMap(data.schema);
    return await this.getIterate<ISchemaMapValue, ISchemaUnitMap>(args, { container: createSchemaUnitMap(), child_ids: Array.from(schema_map.keys()), child_type: "collection", multiple }, (name) =>
      schema_map.get(name) as ISchemaMapValue, (_, { schema_id, name, type }, schema_unit_map) => {
        const schema_obj = new SchemaUnit({ ...this.getProps(), id: this.id, schema_id });
        schema_unit_map[type].set(schema_id, schema_obj)
        schema_unit_map[type].set(name, schema_obj)
      })
  }

  /**
   * Update and return a single column from the collection schema
   * @param args schema_id string and schema properties tuple
   * @returns A SchemaUnit object representing the column
   */
  async updateSchemaUnit(arg: UpdateType<ISchemaMapValue, Partial<TSchemaUnit>>) {
    return (await this.updateSchemaUnits(typeof arg === "function" ? arg : [arg], false))
  }

  /**
   * Update and return multiple columns from the collection schema
   * @param args schema_id string and schema properties array of tuples
   * @returns An array of SchemaUnit objects representing the columns
   */
  async updateSchemaUnits(args: UpdateTypes<ISchemaMapValue, Partial<TSchemaUnit>>, multiple?: boolean) {
    const data = this.getCachedData(), schema_map = getSchemaMap(data.schema);
    const results = await this.updateIterate<ISchemaMapValue, Partial<TSchemaUnit>, ISchemaUnitMap>(args, {
      child_ids: Array.from(schema_map.keys()),
      child_type: "collection",
      multiple,
      manual: true,
      container: createSchemaUnitMap()
    }, (name) => schema_map.get(name) as ISchemaMapValue,
      (_, { schema_id, type, name }, updated_data, results) => {
      data.schema[schema_id] = { ...data.schema[schema_id], ...updated_data } as TSchemaUnit;
      type = updated_data.type ?? type;
      const schema_obj = new SchemaUnit({ schema_id, ...this.getProps(), id: this.id })
      results[type].set(schema_id, schema_obj)
      results[type].set(name, schema_obj)
    });
    this.updateLastEditedProps();
    this.stack.push(Operation.collection.update(this.id, [], { schema: data.schema }))
    return results;
  }

  /**
   * Delete a single column from the collection schema
   * @param args schema_id string or predicate function
   * @returns A SchemaUnit object representing the column
   */
  async deleteSchemaUnit(args?: FilterType<ISchemaMapValue>) {
    return (await this.deleteSchemaUnits(typeof args === "string" ? [args] : args, false));
  }

  /**
   * Delete multiple columns from the collection schema
   * @param args schema_id string array or predicate function
   * @returns An array of SchemaUnit objects representing the columns
   */
  async deleteSchemaUnits(args?: FilterTypes<ISchemaMapValue>, multiple?: boolean) {
    const data = this.getCachedData(), schema_map = getSchemaMap(data.schema);
    await this.deleteIterate<ISchemaMapValue>(args, {
      child_ids: Array.from(schema_map.keys()),
      child_type: "collection",
      multiple,
      manual: true,
      container: []
    }, (name) => {
      const schema_unit_data = schema_map.get(name) as ISchemaMapValue;
      return { ...schema_unit_data, schema_id: schema_unit_data.schema_id };
    }, (_, { schema_id }) => {
      delete data.schema[schema_id]
    });
    this.updateLastEditedProps();
    this.stack.push(Operation.collection.update(this.id, [], { schema: data.schema }))
  }
}

export default Collection;