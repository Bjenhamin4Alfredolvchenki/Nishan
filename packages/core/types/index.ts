export * from './block';
export * from './maps';
export * from './nishan';
