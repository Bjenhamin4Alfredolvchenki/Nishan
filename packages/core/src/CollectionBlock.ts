import Collection from './Collection';
import { TableView, GalleryView, ListView, BoardView, TimelineView, CalendarView } from './View';

import { createViewMap, createViews, transformToMultiple } from '../utils';
import { ICollection, TView, TViewUpdateInput, ICollectionViewPage, ICollectionView } from '@nishans/types';
import {
	NishanArg,
	FilterTypes,
	UpdateType,
	UpdateTypes,
	FilterType,
	TViewCreateInput,
	IViewMap,
	TCollectionBlockInput
} from '../types';
import Block from './Block';
import { Operation } from '@nishans/operations';

const view_class = {
	board: BoardView,
	gallery: GalleryView,
	list: ListView,
	timeline: TimelineView,
	table: TableView,
	calendar: CalendarView
};

/**
 * A class to represent collectionblock type in Notion
 * @noInheritDoc
 */
class CollectionBlock<T extends ICollectionViewPage | ICollectionView> extends Block<T, TCollectionBlockInput> {
	constructor (arg: NishanArg & { type: 'block' }) {
		super({ ...arg });
	}

	/**
   * Fetch the corresponding collection of the collection block using the collection_id
   * @returns The corresponding collection object
   */
	async getCollection () {
		return new Collection({
			...this.getProps(),
			id: this.getCachedData().collection_id
		});
	}

	createViews (params: TViewCreateInput[]) {
		const data = this.getCachedData(),
			collection = this.cache.collection.get(data.collection_id) as ICollection,
			[ view_ids, view_map ] = createViews(collection, params, this.getProps());
		this.Operations.stack.push(Operation.block.update(data.id, [], { view_ids: [ ...data.view_ids, ...view_ids ] }));
		data.view_ids = [ ...data.view_ids, ...view_ids ];
		this.updateLastEditedProps();
		return view_map;
	}

	/**
   * Get all the views associated with the collection block
   * @returns An array of view objects of the collectionblock
   */
	async getViews (args?: FilterTypes<TView>, multiple?: boolean) {
		return await this.getIterate<TView, IViewMap>(
			args,
			{ multiple, container: createViewMap(), child_ids: 'view_ids', child_type: 'collection_view' },
			(view_id) => this.cache.collection_view.get(view_id) as TView,
			(view_id, { type, name }, view_map) => {
				const view_obj = new view_class[type]({
					id: view_id,
					...this.getProps()
				}) as any;
				view_map[type].set(view_id, view_obj);
				view_map[type].set(name, view_obj);
			}
		);
	}

	async updateView (arg: UpdateType<TView, TViewUpdateInput>) {
		return await this.updateViews(transformToMultiple(arg), false);
	}

	async updateViews (args: UpdateTypes<TView, TViewUpdateInput>, multiple?: boolean) {
		return await this.updateIterate<TView, TViewUpdateInput, IViewMap>(
			args,
			{
				container: createViewMap(),
				multiple,
				child_ids: 'view_ids',
				child_type: 'collection_view'
			},
			(view_id) => this.cache.collection_view.get(view_id),
			(id, { type, name }, _, view_map) => {
				const view_obj = new view_class[type]({ ...this.getProps(), id }) as any;
				view_map[type].set(id, view_obj);
				view_map[type].set(name, view_obj);
			}
		);
	}

	/**
   * Delete a single root page from the space
   * @param arg Criteria to filter the page to be deleted
   */
	async deleteView (arg?: FilterType<TView>) {
		return await this.deleteViews(transformToMultiple(arg), false);
	}

	/**
   * Delete multiple root_pages or root_collection_view_pages
   * @param arg Criteria to filter the pages to be deleted
   * @param multiple whether or not multiple root pages should be deleted
   */
	async deleteViews (args?: FilterTypes<TView>, multiple?: boolean) {
		await this.deleteIterate<TView>(
			args,
			{
				child_ids: 'view_ids',
				child_path: 'view_ids',
				child_type: 'collection_view',
				multiple,
				container: []
			},
			(view_id) => this.cache.collection_view.get(view_id)
		);
	}
}

export default CollectionBlock;
