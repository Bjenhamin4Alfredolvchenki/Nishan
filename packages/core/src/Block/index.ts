import CollectionViewPage from './CollectionViewPage';
import CollectionView from './CollectionView';
import Page from './Page';
import CollectionBlock from './CollectionBlock';
import Block from './Block';

export { Block, CollectionViewPage, CollectionView, Page, CollectionBlock };
