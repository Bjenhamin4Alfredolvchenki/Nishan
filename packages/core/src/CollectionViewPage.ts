import { ISpace, IPage, ICollectionViewPage } from '@nishans/types';
import { NotionPermissions } from '../src';
import { NishanArg } from '../types';
import CollectionBlock from './CollectionBlock';

/**
 * A class to represent collectionviewpage of Notion
 * @noInheritDoc
 */
class CollectionViewPage extends CollectionBlock<ICollectionViewPage> {
	Permissions: NotionPermissions;
	constructor (arg: NishanArg) {
		super({ ...arg });
		this.Permissions = new NotionPermissions(arg, arg.id, 'block');
	}

	getCachedParentData () {
		const data = this.getCachedData();
		return this.cache[data.parent_table].get(data.parent_id) as IPage | ISpace;
	}
}

export default CollectionViewPage;
