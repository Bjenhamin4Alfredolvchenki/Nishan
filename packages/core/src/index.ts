import {NotionCacheClass} from "@nishans/cache";
import { INotionUser } from "@nishans/types";

import Block from "./Block";
import Collection from "./Collection";
import CollectionViewPage from "./CollectionViewPage";
import CollectionView from "./CollectionView";
import NotionUser from "./NotionUser";
import Page from "./Page";
import UserSettings from "./UserSettings";
import UserRoot from "./UserRoot";
import SpaceView from "./SpaceView";
import Space from "./Space";
import SchemaUnit from "./SchemaUnit";
import CollectionBlock from "./CollectionBlock";
import NotionPermissions from "./Permissions";
export * from "./View";
import { constructLogger, error, iterateAndGetChildren, transformToMultiple } from "../utils";
import {Logger, NishanArg,FilterType, FilterTypes} from "../types";

class Nishan extends NotionCacheClass {
  token: string;
  interval: number;
  init_cache: boolean;
  logger: Logger;

  constructor(arg: Pick<NishanArg, "token"> & { interval?: number, logger?: Logger }) {
    super(arg);
    this.token = arg.token;
    this.interval = arg.interval ?? 500;
    this.init_cache = false;
    this.logger = constructLogger(arg.logger)
  }

  #initializeCache = async () => {
    if (!this.init_cache) {
      await this.initializeNotionCache();
      this.init_cache = true;
    }
  }

  /**
   * Get `INotionUser` and return `NotionUser` by the `args` param
   * @param args An string id, a predicate passed the INotionUser or undefined to indicate everything
   */
  async getNotionUser(arg?: FilterType<INotionUser>) {
    return (await this.getNotionUsers(transformToMultiple(arg), false))[0];
  }

  /**
   * Get `INotionUser[]` and return `NotionUser[]` by the `args` param
   * @param args An array of string ids, a predicate passed the INotionUser or undefined to indicate everything
   */
  async getNotionUsers(args?: FilterTypes<INotionUser>, multiple?: boolean) {
    await this.#initializeCache();

    const common_props = {
      token: this.token,
      cache: this.cache,
      interval: this.interval,
      logger: this.logger,
      stack: [],
    }

    const notion_user_ids: string[] = [];
    for (const [notion_user_id] of this.cache.notion_user) {
      notion_user_ids.push(notion_user_id)
    }

    return await iterateAndGetChildren<INotionUser, INotionUser, NotionUser[]>(args, (id)=>this.cache.notion_user.get(id), {
      ...common_props,
      multiple,
      child_ids: notion_user_ids,
      child_type: 'notion_user',
      container: [],
      parent_id: '',
      parent_type: 'notion_user'
    }, (id, _, container)=>container.push(new NotionUser({ ...common_props, user_id: id, id, space_id: "0", shard_id: 0 })));
  }
}

export default Nishan;
export * from "../types";
export * from "../utils";

export {
  Block,
  Collection,
  CollectionViewPage,
  CollectionView,
  NotionUser,
  Page,
  UserSettings,
  UserRoot,
  SpaceView,
  Space,
  SchemaUnit,
  CollectionBlock,
  NotionPermissions
}