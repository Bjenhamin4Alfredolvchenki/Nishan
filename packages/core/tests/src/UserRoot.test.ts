import { ICache } from '@nishans/cache';
import { IOperation } from '@nishans/types';
import deepEqual from 'deep-equal';
import { UserRoot } from '../../src';

afterEach(() => {
	jest.restoreAllMocks();
});

it(`get space_views`, async () => {
	const cache: ICache = {
			block: new Map(),
			collection: new Map(),
			collection_view: new Map(),
			notion_user: new Map(),
			space: new Map(),
			space_view: new Map([ [ 'space_view_1', { alive: true } as any ] ]),
			user_root: new Map([ [ 'user_root_1', { space_views: [ 'space_view_1' ] } as any ] ]),
			user_settings: new Map()
		},
		stack: IOperation[] = [];

	const user_root = new UserRoot({
		cache,
		id: 'user_root_1',
		stack,
		interval: 0,
		shard_id: 123,
		space_id: 'space_1',
		token: 'token',
		user_id: 'user_root_1'
	});

	const space_view = await user_root.getSpaceView('space_view_1');
	expect(deepEqual(space_view.getCachedData(), { alive: true })).toBe(true);
});

it(`update space_views`, async () => {
	const cache: ICache = {
			block: new Map(),
			collection: new Map(),
			collection_view: new Map(),
			notion_user: new Map(),
			space: new Map(),
			space_view: new Map([ [ 'space_view_1', { alive: true } as any ] ]),
			user_root: new Map([ [ 'user_root_1', { space_views: [ 'space_view_1' ] } as any ] ]),
			user_settings: new Map()
		},
		stack: IOperation[] = [];

	const user_root = new UserRoot({
		cache,
		id: 'user_root_1',
		stack,
		interval: 0,
		shard_id: 123,
		space_id: 'space_1',
		token: 'token',
		user_id: 'user_root_1'
	});

	const space_view = await user_root.updateSpaceView([ 'space_view_1', { joined: false } ]);
	expect(space_view.getCachedData()).toMatchSnapshot({
		alive: true,
		last_edited_time: expect.any(Number),
		last_edited_by_table: 'notion_user',
		last_edited_by_id: 'user_root_1',
		joined: false
	} as any);

	expect(stack).toMatchSnapshot([
		{
			table: 'space_view',
			id: 'space_view_1',
			command: 'update',
			path: [],
			args: {
				last_edited_time: expect.any(Number),
				last_edited_by_table: 'notion_user',
				last_edited_by_id: 'user_root_1',
				joined: false
			}
		},
		{
			table: 'user_root',
			id: 'user_root_1',
			command: 'update',
			path: [],
			args: {
				last_edited_time: expect.any(Number),
				last_edited_by_table: 'notion_user',
				last_edited_by_id: 'user_root_1'
			}
		}
	] as any);
});
