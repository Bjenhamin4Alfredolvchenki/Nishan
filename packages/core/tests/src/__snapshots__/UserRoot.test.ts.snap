// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update space_views 1`] = `
Object {
  "alive": true,
  "joined": false,
  "last_edited_by_id": "user_root_1",
  "last_edited_by_table": "notion_user",
  "last_edited_time": Any<Number>,
}
`;

exports[`update space_views 2`] = `
Array [
  Object {
    "args": Object {
      "joined": false,
      "last_edited_by_id": "user_root_1",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
    },
    "command": "update",
    "id": "space_view_1",
    "path": Array [],
    "table": "space_view",
  },
  Object {
    "args": Object {
      "last_edited_by_id": "user_root_1",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
    },
    "command": "update",
    "id": "user_root_1",
    "path": Array [],
    "table": "user_root",
  },
]
`;
